Add pebble bench flag to find peak sustainable throughput

ajkr opened the pull request:
Peak sustainable throughput is the max possible throughput not leading
to stalls. We estimate it by using binary search to find the highest
throughput where stalls are not seen and background work backlog is
non-increasing over a period of time.

benchmark setup (on c5d.xlarge):
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload insert=100 --values 64 --initial-keys 0 --keys uniform --concurrency 64 --batch 64 --num-ops $[(4 << 30) / 64 / 64] --duration 0s --wait-compactions
```

optionally add initial compaction debt:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --verbose
```

benchmark command:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --find-peak-ops-per-sec --verbose
```

results:

initial debt (GB) | trial | peak keys/sec | final debt (GB)
-- | -- | -- | --
0 | 1 | 68992 | 7
0 | 2 | 70272 | 7
16 | 1 | 101824 | 21
16 | 2 | 85312 | 13

ajkr committed:
Add pebble bench flag to find peak sustainable throughput

Peak sustainable throughput is the max possible throughput not leading
to stalls. We estimate it by using binary search to find the highest
throughput where stalls are not seen and background work backlog is
non-increasing over a period of time.

benchmark setup (on c5d.xlarge):
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload insert=100 --values 64 --initial-keys 0 --keys uniform --concurrency 64 --batch 64 --num-ops $[(4 << 30) / 64 / 64] --duration 0s --wait-compactions
```

optionally add initial compaction debt:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --verbose
```

benchmark command:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --find-peak-ops-per-sec --verbose
```

results:

initial debt (GB) | trial | peak keys/sec | final debt (GB)
-- | -- | -- | --
0 | 1 | 68992 | 7
0 | 2 | 70272 | 7
16 | 1 | 101824 | 21
16 | 2 | 85312 | 13

petermattis commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/petermattis/pebble/241)
<!-- Reviewable:end -->

ajkr commented:
One possible way to show it's good enough would be to compare the overwrite benchmark result on a DB that's initially fully compacted vs. one that's initially near the stall limit. Currently the gap it shows in that scenario is too huge, so [wip] for now.

petermattis reviewed (commented):
Note that we configure RocksDB with a very high L0 stop writes threshold (400!). The configuration for Pebble is much lower (32). We should probably fix that discrepancy so we're comparing fairly between the two.

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LmuFXaW3wPoXAdnBn7u:byu6lxp)* status: 0 of 9 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

ajkr reviewed (commented):
Unfortunately I'm quite far from the point it'll be useful to compare two systems (or even one system for that matter). But I'll keep that in mind and probably go with the lower value (32) once we eventually get there!

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LmuMxxPBZ86SSVd_v4q:b-tdbn3j)* status: 0 of 9 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

Ryanfsdf reviewed (dismissed):
Just chiming in since I thought this PR was interesting.

Left a few comments regarding compaction debt!

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LmuPqz34T9P6YB0yO16:b-9umjsi)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @ajkr)

___
*[db.go, line 811 at r1](https://reviewable.io/reviews/petermattis/pebble/241#-LmuKuAJBdB2jDVqIMnH:-LmuKuAJBdB2jDVqIMnI:b-n6szf9) ([raw file](https://github.com/petermattis/pebble/blob/90956adfe3e10640348850d7530cf6f7caa2d1b1/db.go#L811)):*
> ```Go
> 			metrics.Levels[level].Score = float64(metrics.Levels[level].Size) / float64(p.levelMaxBytes[level])
> 		}
> 		metrics.EstimatedCompactionDebt = p.estimatedCompactionDebt(0 /*l0ExtraSize*/)
> ```

I think if you do
```
bytesFlushed := atomic.LoadUint64(&d.bytesFlushed)
bytesCompacted := atomic.LoadUint64(&d.bytesCompacted)
metrics.EstimatedCompactionDebt = p.estimatedCompactionDebt(bytesFlushed) - bytesCompacted
```
it should account for the current flush and compaction, which should give a smoother compaction debt estimation.

Then the doc for `bytesCompacted` can be updated from 
```
// bytesCompacted is the number of bytes compacted in the current compaction.
// This is used as a dummy variable to increment during compaction, and the
// value is not used anywhere.
```
to something like
```
// bytesCompacted is the number of bytes compacted in the current compaction.
// This is incremented during compaction, and the value is used for metrics. It must
// be read/written atomically.
```

I should have added some more documentation to `estimatedCompactionDebt` to make this clearer.
___
*[cmd/pebble/test.go, line 309 at r1](https://reviewable.io/reviews/petermattis/pebble/241#-LmuMbLw3NuZbpkGe7qR:-LmuMbLw3NuZbpkGe7qS:b-kl73p9) ([raw file](https://github.com/petermattis/pebble/blob/90956adfe3e10640348850d7530cf6f7caa2d1b1/cmd/pebble/test.go#L309)):*
> ```Go
> 		//     of a fixed number of intervals, we could require a trend lasts for one (or
> 		//     more) full L0->Lbase compaction cycles.
> 		// (3) at low compaction debt levels (below compactionSlowdownThreshold), the
> ```

I think it might work better to keep the rate the same if the compaction debt is well below `compactionSlowdownThreshold` (Less than 80% of `compactionSlowdownThreshold` or something like that). So basically, ignoring the compaction debt metrics when compaction debt is far below the threshold. When the database is running at a stable state, the compaction debt should typically be +/- 5% of `compactionSlowdownThreshold`. I think there should only be a few instances where the compaction debt metrics would be ignored, which I assume would be during startup.

`compactionSlowdownThreshold` is essentially the "zero" level of compaction debt, and it's calculated by as the maximum amount of compaction debt that can be incurred by a single flush. You can think of this as the "105% memtable size" threshold that we use for flush pacing. I'm hoping this value isn't large enough to visibly impact the throughput.


<!-- Sent from Reviewable.io -->

ajkr reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @Ryanfsdf)

___
*[db.go, line 811 at r1](https://reviewable.io/reviews/petermattis/pebble/241#-LmuKuAJBdB2jDVqIMnH:-LmuRdHF9Y3YBc0hc0uW:bdhnoeo) ([raw file](https://github.com/petermattis/pebble/blob/90956adfe3e10640348850d7530cf6f7caa2d1b1/db.go#L811)):*
<details><summary><i>Previously, Ryanfsdf (Ryan Kim) wrote…</i></summary><blockquote>

I think if you do
```
bytesFlushed := atomic.LoadUint64(&d.bytesFlushed)
bytesCompacted := atomic.LoadUint64(&d.bytesCompacted)
metrics.EstimatedCompactionDebt = p.estimatedCompactionDebt(bytesFlushed) - bytesCompacted
```
it should account for the current flush and compaction, which should give a smoother compaction debt estimation.

Then the doc for `bytesCompacted` can be updated from 
```
// bytesCompacted is the number of bytes compacted in the current compaction.
// This is used as a dummy variable to increment during compaction, and the
// value is not used anywhere.
```
to something like
```
// bytesCompacted is the number of bytes compacted in the current compaction.
// This is incremented during compaction, and the value is used for metrics. It must
// be read/written atomically.
```

I should have added some more documentation to `estimatedCompactionDebt` to make this clearer.
</blockquote></details>

Oh that's perfect, thanks! I didn't realize the accounting for pending flush/compaction bytes is already done.
___
*[cmd/pebble/test.go, line 309 at r1](https://reviewable.io/reviews/petermattis/pebble/241#-LmuMbLw3NuZbpkGe7qR:-LmuSAyP8S-j6b5DcwbO:b-mxpn22) ([raw file](https://github.com/petermattis/pebble/blob/90956adfe3e10640348850d7530cf6f7caa2d1b1/cmd/pebble/test.go#L309)):*
<details><summary><i>Previously, Ryanfsdf (Ryan Kim) wrote…</i></summary><blockquote>

I think it might work better to keep the rate the same if the compaction debt is well below `compactionSlowdownThreshold` (Less than 80% of `compactionSlowdownThreshold` or something like that). So basically, ignoring the compaction debt metrics when compaction debt is far below the threshold. When the database is running at a stable state, the compaction debt should typically be +/- 5% of `compactionSlowdownThreshold`. I think there should only be a few instances where the compaction debt metrics would be ignored, which I assume would be during startup.

`compactionSlowdownThreshold` is essentially the "zero" level of compaction debt, and it's calculated by maximum amount of compaction debt that can be incurred by a single flush. You can think of this as the "105% memtable size" threshold that we use for flush pacing. I'm hoping this value isn't large enough to visibly impact the throughput.

</blockquote></details>

Oh I actually ended up treating compaction debt below 1GB as "non-increasing" (see https://github.com/petermattis/pebble/pull/241/files#diff-86b4755792f30492ab682e1d5a11cc28R365-R371). So if we see that for a while we'll increase the rate. I can update this TODO to make it clear that we at least attempted to solve the problem, although it is not perfect.


<!-- Sent from Reviewable.io -->

Ryanfsdf reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @ajkr and @Ryanfsdf)

___
*[cmd/pebble/test.go, line 309 at r1](https://reviewable.io/reviews/petermattis/pebble/241#-LmuMbLw3NuZbpkGe7qR:-LmuUMQv4Z8Qwe1q4Eif:b-n15k7w) ([raw file](https://github.com/petermattis/pebble/blob/90956adfe3e10640348850d7530cf6f7caa2d1b1/cmd/pebble/test.go#L309)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

Oh I actually ended up treating compaction debt below 1GB as "decreasing" (see https://github.com/petermattis/pebble/pull/241/files#diff-86b4755792f30492ab682e1d5a11cc28R365-R371). So if we see that for a while we'll increase the rate. I can update this TODO to make it clear that we at least attempted to solve the problem, although it is not perfect.
</blockquote></details>

Oh I see. That would work too!


<!-- Sent from Reviewable.io -->

ajkr commented:
OK, addressed comments to use smoothed compaction debt and also fixed an underflow bug that was causing us to sometimes change the rate in the wrong direction. Upon fixing those I'm finding pretty repeatable results for Pebble.

To test I created a small-ish DB (4GB) and ran overwrite benchmarks on it.

benchmark setup (on c5d.xlarge):
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload insert=100 --values 64 --initial-keys 0 --keys uniform --concurrency 64 --batch 64 --num-ops $[(4 << 30) / 64 / 64] --duration 0s --wait-compactions
```

optionally add compaction debt:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --verbose
```

benchmark command:
```
$ ./pebble bench ycsb /mnt/data1/bench_16G/ --workload update=100 --concurrency 64 --batch 64 --values 64 --prepopulated-keys $[(4 << 30) / 64] --initial-keys 0 --keys uniform --duration 0s --find-peak-ops-per-sec --verbose
```

results:

initial debt (GB) | trial | keys/sec | final debt (GB)
-- | -- | -- | --
0 | 1 | 68992 | 7
0 | 2 | 70272 | 7
16 | 1 | 101824 | 21
16 | 2 | 85312 | 13


The peak keys/sec appears highly correlated with the final compaction debt. This is aligned with expectations set in the TODO labeled (3) in the code. I think this flag is good enough now to be useful for Pebble, like to monitor regression or measure improvement of write efficiency optimizations. We just have to be careful to only make comparisons between runs with similar final compaction debts.

The next steps are solving TODO (3) and supporting RocksDB. They will be in future PRs if I am able to solve them.

petermattis reviewed (commented):
I'll give this a full look tomorrow. My first impression is that the `find-peek-ops-per-sec` stuff is quite intrusive into the main test loop. Is there a way to factor that out? Does it need to be part of the main test loop, or could we have a separate goroutine for the find-peek-ops loop?

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-Lmvtffw0r7LcFKWK-iI:bbqk2pp)* status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @ajkr, @petermattis, and @Ryanfsdf)



<!-- Sent from Reviewable.io -->

ajkr commented:
Yes, it could be another goroutine. Then we can do our own ticking and call `db.Metrics()` at whatever interval we want. Currently we piggyback on the call to `db.Metrics()` used for logging, which is called every ten seconds. It works well enough but, who knows, maybe separating it and using a lower interval would allow us to get similarly reliable results with a shorter benchmark time. Right now it's ~20 minutes which is pretty long.

petermattis reviewed (commented):
The general idea looks good to me, though as I mentioned previously, I'd like to see this encapsulated.

How stable are the results from run to run?

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-Lmzc1LaAsnUuZgEfncK:bmvxbie)* status: 0 of 10 files reviewed, 6 unresolved discussions (waiting on @ajkr and @Ryanfsdf)

___
*[cmd/pebble/test.go, line 301 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-LmzYrFE4lyUcRlcTZiV:-LmzYrFF9e1TYjSFqut0:b-jyailx) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L301)):*
> ```Go
> 		// we adjusted the workload rate.
> 		numIntervals int
> 	}
> ```

I think you should pull this out into a top-level struct in its own file. And let's create a separate metrics loop for the peak-rate finder. 
___
*[cmd/pebble/test.go, line 385 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-Lmz_Eke2f5KycjSezCr:-Lmz_Eke2f5KycjSezCs:b-c86uj4) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L385)):*
> ```Go
> 							findPeakInfo.increased = increased
> 							findPeakInfo.numSameIncreased = 1
> 						}
> ```

The current approach is tracking a trend of collection periods where the compaction debt is increasing/decreasing. The state feels complicated as I read this. I wonder if it would be clearer to track a history of the last N compaction debt measurements and provide a function that loops over those measurements and decides what to do. This history would be cleared every time the rate is adjusted. I'm imagining state like:

```
compactionDebtHistory []uint64
```

This would replace the `prevCompactionDebt`, `increased`, `numSameIncreased`, and `numIntervals` state.
___
*[cmd/pebble/test.go, line 388 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-Lmzab-T0ND8ZWK4H_BI:-Lmzab-U4P005Xklp1LL:b-lhoz66) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L388)):*
> ```Go
> 						findPeakInfo.numIntervals++
> 						intervalRate := float64(limiter.TotalReserved()-findPeakInfo.prevTotalReserved) /
> 							time.Since(findPeakInfo.prevUpdated).Seconds()
> ```

So this is calculating the ops/sec by using the rate limiter count? I would have thought there is a more direct measurement we could get from the histograms or workloads, but looking I don't see anything. You could sum over the histograms, though that seems awkward. I'm only bringing this up because using the rate limiter to measure the rate feels odd to me.
___
*[cmd/pebble/test.go, line 431 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-LmzbbLeEwsajx97cYx5:-LmzbbLeEwsajx97cYx6:b8i1km3) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L431)):*
> ```Go
> 		case <-findPeakInfo.stallEnded:
> 			// Now that the stall ended, try a new (reduced) rate
> 			adjustRate()
> ```

I wonder if we should wait for background compactions to complete before starting up the workload again. This could take much longer, but might lead to a more stable answer.


<!-- Sent from Reviewable.io -->

ajkr commented:
TFTR.

> How stable are the results from run to run?

In the PR description there are sample results from a few runs. My conclusion is it's pretty stable as long as the benchmarks end with a similar amount of compaction debt. It's not surprising I think that the max rate that causes non-increasing debt is higher at higher debt levels due to the lower fanout between levels.

I saw you mentioned perhaps waiting for background compaction to finish before trying a new rate. That'd certainly solve stability as debt at the end of the benchmark would always be pretty small. I am hoping to go the other way, though, and ensure compaction debt is near the limit before trying a new rate. That'd achieve stability as debt at the end of the benchmark would always be near the limit, and also give us the highest possible number (the true "peak" sustainable throughput) as we know write-amp is minimal at that debt level.

petermattis commented:
> I saw you mentioned perhaps waiting for background compaction to finish before trying a new rate. That'd certainly solve stability as debt at the end of the benchmark would always be pretty small. I am hoping to go the other way, though, and ensure compaction debt is near the limit before trying a new rate. That'd achieve stability as debt at the end of the benchmark would always be near the limit, and also give us the highest possible number (the true "peak" sustainable throughput) as we know write-amp is minimal at that debt level.

My suggestion was also aimed at making sure we had stability at each phase of the test. If you target rate limit Y and see a write stall, simply starting up load again (at rate limit X < Y) as soon as the write stall end could hit a situation where the detritus from the previous write stall (compaction debt) hasn't had a chance to clear.  I don't have any evidence that this is the case, just a suspicion.

ajkr reviewed (commented):
Ideally we'd have the same starting point each time we try a different rate, whether it's a low debt starting point or a high one. In RocksDB's perf regression, each benchmark runs on its own checkpoint of a single pre-generated DB, which ensures the starting point is the same across benchmark runs. I know live checkpoints aren't available in Pebble (though maybe we'll need to add them since cockroach uses them), so for now we can make offline checkpoints by just hardlinking SSTs and copying all other files. And by "for now" I really mean in a future PR.

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LnFcIce-MrGj7tedhQE:b-wjcteo)* status: 0 of 10 files reviewed, 5 unresolved discussions (waiting on @petermattis and @Ryanfsdf)

___
*[cmd/pebble/test.go, line 301 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-LmzYrFE4lyUcRlcTZiV:-LnKQSmBCf9PXTaJOqDG:b-896fix) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L301)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think you should pull this out into a top-level struct in its own file. And let's create a separate metrics loop for the peak-rate finder. 
</blockquote></details>

Done.
___
*[cmd/pebble/test.go, line 385 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-Lmz_Eke2f5KycjSezCr:-LnKS9ZB16EB26KeALM3:b-m3z7rt) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L385)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The current approach is tracking a trend of collection periods where the compaction debt is increasing/decreasing. The state feels complicated as I read this. I wonder if it would be clearer to track a history of the last N compaction debt measurements and provide a function that loops over those measurements and decides what to do. This history would be cleared every time the rate is adjusted. I'm imagining state like:

```
compactionDebtHistory []uint64
```

This would replace the `prevCompactionDebt`, `increased`, `numSameIncreased`, and `numIntervals` state.
</blockquote></details>

Will try it out.
___
*[cmd/pebble/test.go, line 388 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-Lmzab-T0ND8ZWK4H_BI:-LnKQbwhAY68Erksj2Qv:b-dsawvz) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L388)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

So this is calculating the ops/sec by using the rate limiter count? I would have thought there is a more direct measurement we could get from the histograms or workloads, but looking I don't see anything. You could sum over the histograms, though that seems awkward. I'm only bringing this up because using the rate limiter to measure the rate feels odd to me.
</blockquote></details>

I also didn't notice another convenient way to do it. It looked like each test manages its own histograms and so it might be non-trivial to move ownership to `runTest()`, but I can look again. OTOH there's nothing requiring the histogram counts exactly match the number of ops requested from rate limiter, while the latter is what we really want, so maybe getting it directly is fine.
___
*[cmd/pebble/test.go, line 431 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-LmzbbLeEwsajx97cYx5:-LnKRrCi7dunWof9tAQD:b1qhy2k) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L431)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if we should wait for background compactions to complete before starting up the workload again. This could take much longer, but might lead to a more stable answer.
</blockquote></details>

Replied in the review discussion.


<!-- Sent from Reviewable.io -->

ajkr reviewed (commented):
> My suggestion was also aimed at making sure we had stability at each phase of the test. If you target rate limit Y and see a write stall, simply starting up load again (at rate limit X < Y) as soon as the write stall end could hit a situation where the detritus from the previous write stall (compaction debt) hasn't had a chance to clear.  I don't have any evidence that this is the case, just a suspicion.

OK, thought through this again. I think you are right. If we just recovered from a write stall our levels will have higher compaction scores, which means in the next trial we will start L0->Lbase later and it will take longer. So we will be more likely to hit the L0 file count limit than we would have been if we had begun in a state where compaction was all caught up.

Without resorting to a clean slate for every rate trial, maybe we can forgive write stalls if compaction debt is decreasing in the intervals before/after the stall. Here, forgiveness means not adjusting the rate and simply continuing on with the trial.

>  I am hoping to go the other way, though, and ensure compaction debt is near the limit before trying a new rate. That'd achieve stability as debt at the end of the benchmark would always be near the limit, and also give us the highest possible number (the true "peak" sustainable throughput) as we know write-amp is minimal at that debt level.

To clarify my quoted comment, this idea would work if disk space is the only stall trigger. It does not work for us though since L0 file count limit is also important.

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LnPaJNXA2LdCq142_SS:b-h406yo)* status: 0 of 11 files reviewed, 5 unresolved discussions (waiting on @petermattis and @Ryanfsdf)

___
*[cmd/pebble/test.go, line 385 at r2](https://reviewable.io/reviews/petermattis/pebble/241#-Lmz_Eke2f5KycjSezCr:-LnPa8U64LX8o8fNqoRz:b-896fix) ([raw file](https://github.com/petermattis/pebble/blob/145b0e54e8d82b7a7de97bc7f37fabd941aaf0a5/cmd/pebble/test.go#L385)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

Will try it out.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
The encapsulation looks good.

> OK, thought through this again. I think you are right. If we just recovered from a write stall our levels will have higher compaction scores, which means in the next trial we will start L0->Lbase later and it will take longer. So we will be more likely to hit the L0 file count limit than we would have been if we had begun in a state where compaction was all caught up.
> 
> Without resorting to a clean slate for every rate trial, maybe we can forgive write stalls if compaction debt is decreasing in the intervals before/after the stall. Here, forgiveness means not adjusting the rate and simply continuing on with the trial.

Another thought is to compact the entire DB when a write stall is seen (e.g. `DB.Compact(nil, nil)`). This isn't quite starting from a clean slate, but reasonably close. I wonder if starting from a completely clean slate would be bad (i.e. too slow). That would require a restructuring of how this works so that the DB can be torn down and recreated. The `DB.Compact` approach would definitely be simpler. Or perhaps your idea of forgiving write stalls could work. I think experimentation will be necessary here.

> *[Reviewable](https://reviewable.io/reviews/petermattis/pebble/241#-:-LncSWi4EIX7dp8hV19h:b-9b3iia)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @ajkr and @Ryanfsdf)

___
*[metrics.go, line 88 at r3](https://reviewable.io/reviews/petermattis/pebble/241#-LncRVayC1nV8JPdpVy-:-LncRVayC1nV8JPdpVy0:b-u7eslr) ([raw file](https://github.com/petermattis/pebble/blob/aa3763a8c5611dc3b2a98f720a307f0aaacf1cad/metrics.go#L88)):*
> ```Go
> 	}
> 	Levels                  [numLevels]LevelMetrics
> 	EstimatedCompactionDebt uint64
> ```

Should we add this to the `VersionMetrics.String()` output? Fine for this to be a TODO for a future PR.
___
*[cmd/pebble/peak_rate.go, line 54 at r3](https://reviewable.io/reviews/petermattis/pebble/241#-LncS4AcDt7zi2ieHNYG:-LncS4AcDt7zi2ieHNYH:b91ektm) ([raw file](https://github.com/petermattis/pebble/blob/aa3763a8c5611dc3b2a98f720a307f0aaacf1cad/cmd/pebble/peak_rate.go#L54)):*
> ```Go
> 	)
> 
> 	// The compaction debts seen in the current rate trial
> ```

Nit: terminate comments with a period. The only exception to this (for Go-style) is single-line comments that appear on the same line as code. `golint` or one of the other linters would complain about this. I suppose we should get that set up to be run be CI automatically.
___
*[cmd/pebble/test.go, line 318 at r3](https://reviewable.io/reviews/petermattis/pebble/241#-LncQtXH3thBLvR7he6P:-LncQtXI802XS5DfKvMO:b-nbijz) ([raw file](https://github.com/petermattis/pebble/blob/aa3763a8c5611dc3b2a98f720a307f0aaacf1cad/cmd/pebble/test.go#L318)):*
> ```Go
> 				log.Fatalf("unable to achieve rate in range [%.f, %.f]; peak sustainable throughput not found\n",
> 					bounds.lower, bounds.upper)
> 			} else if bounds.upper-bounds.lower < bounds.upper/100 || bounds.upper-bounds.lower < 1 {
> ```

This conditional is complex enough that it could be put in a method on `findPeakBounds`.


<!-- Sent from Reviewable.io -->

xxmplus commented:
Close to due to age
